src/Admin/RESTPreloader.php

```diff
@@ -0,0 +1,79 @@
+<?php
+/**
+ * Class ReaderThemeLoader.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP\Admin;
+
+use AmpProject\AmpWP\Infrastructure\Conditional;
+use AmpProject\AmpWP\Infrastructure\Delayed;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+use WP_Screen;
+
+/**
+ * Preloads REST responses for client-side applications to prevent having to call fetch on pageload.
+ *
+ * @package AmpProject\AmpWP
+ */
+final class RESTPreloader implements Conditional, Delayed, Registerable, Service {
+
+	/**
+	 * Paths to preload.
+	 *
+	 * @var array
+	 */
+	private $paths = [];
+
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed() {
+		return function_exists( 'rest_preload_api_request' ) && is_admin() && has_filter( 'amp_preload_rest_paths' );
+	}
+
+	/**
+	 * Get the action to use for registering the service.
+	 *
+	 * @return string Registration action to use.
+	 */
+	public static function get_registration_action() {
+		return 'admin_enqueue_scripts';
+	}
+
+	/**
+	 * Adds hooks.
+	 */
+	public function register() {
+		$screen = get_current_screen();
+
+		/**
+		 * Filters REST API paths to preload for the current page.
+		 *
+		 * @param array $paths Paths to preload.
+		 * @param string|null $screen_id Current screen ID or null if no current screen is set.
+		 */
+		$this->paths = apply_filters( 'amp_preload_rest_paths', $this->paths, is_a( $screen, WP_Screen::class ) ? $screen->id : null );
```

Comment:
Wondering if this filter should be replaced with something more dependency injection-y. Open to suggestions.

Reply:
Yes, I agree. Inject the `RESTPreloader` service via the constructor of the services that need it, and add a `add_preloaded_path()` method or so that these services can then use.

In this case, having the filter adds an unnecessary public API that we don't want to maintain at this point, as it is only a performance tweak.

Reply:
Done in 54123ee949b8c38539efb71489041d22093e30c4